editoast/src/views/authz.rs

```diff
@@ -83,161 +179,579 @@ async fn whoami(Extension(auth): AuthenticationExt) -> Result<Json<WhoamiRespons
-        .await
-        .map_err(AuthorizationError::from)?
-    {
-        return Err(AuthorizationError::Forbidden.into());
+    Json(body): Json<BodyUpdateGrants>,
+) -> Result<impl IntoResponse> {
+    let authorizer = auth.authorizer()?;
+    if body.grant.is_none() && body.revoke.is_none() {
+        return Err(AuthzError::BadGrantsOperation {
+            message: "No grant and revoke operation".to_string(),
+        }
+        .into());
+    }
+    if body.grant.is_some() && body.revoke.is_some() {
+        return Err(AuthzError::BadGrantsOperation {
+            message: "Grant and Revoke cannot be defined at the same time".to_string(),
+        }
+        .into());
     }
 
-    check_user_exists(user_id, &regulator).await?;
+    if let Some(grants) = body.grant {
+        for grant in grants {
+            if grant.resource_type != Resource::Infra {
+                return Err(ResourceError::ResourceTypeNotFound {
+                    resource_type: grant.resource_type.to_string(),
+                })?;
+            }
+            match grant.grant {
+                InfraGrant::Reader => {
+                    authorizer
+                        .grant_infra_reader(grant.subject_id, grant.resource_id)
+                        .await
+                        .map_err(AuthzError::from)?;
+                }
+                InfraGrant::Writer => {
+                    authorizer
+                        .grant_infra_writer(grant.subject_id, grant.resource_id)
+                        .await
+                        .map_err(AuthzError::from)?;
+                }
+                InfraGrant::Owner => {
+                    authorizer
+                        .grant_infra_owner(grant.subject_id, grant.resource_id)
+                        .await
+                        .map_err(AuthzError::from)?;
+                }
+            }
+        }
+    }
+    if let Some(revoke) = body.revoke {
+        for grant in revoke {
+            if grant.resource_type != Resource::Infra {
+                return Err(ResourceError::ResourceTypeNotFound {
+                    resource_type: grant.resource_type.to_string(),
+                })?;
+            }
+            authorizer
+                .revoke_infra_reader(grant.subject_id, grant.resource_id)
+                .await
+                .map_err(AuthzError::from)?;
+
+            authorizer
+                .revoke_infra_writer(grant.subject_id, grant.resource_id)
+                .await
+                .map_err(AuthzError::from)?;
+
+            authorizer
+                .revoke_infra_owner(grant.subject_id, grant.resource_id)
+                .await
+                .map_err(AuthzError::from)?;
+        }
+    }
 
-    regulator
-        .revoke_user_roles(user_id, HashSet::from_iter(roles))
-        .await
-        .map_err(AuthzError::from)?;
-    Ok(axum::http::StatusCode::NO_CONTENT)
+    Ok(StatusCode::NO_CONTENT)
 }
 
 #[cfg(test)]
-mod tests {
+pub mod test {
```

Comment:
```suggestion
mod tests {
```

no reason to make it pub and it's `tests` by convention